test/filt_order.jl

```diff
@@ -208,8 +208,23 @@ end
 
 # using some simple examples for testing Brent's method.
 @testset "brent" begin
-    f1(x) = (x+3) * ((x-1)^2) # x³ + x² - 5x + 3
+    f1(x) = (x + 3) * ((x - 1)^2) # x³ + x² - 5x + 3
     @test ≈(f1(DSP.Filters.brent(f1, -4.0, 4.0)), 0.0, atol=1e-8)
     @test ≈(sin(DSP.Filters.brent(sin, 0.0, 2pi)), -1.0, atol=1e-8)
     @test ≈(cos(DSP.Filters.brent(cos, 0.0, 2pi)), -1.0, atol=1e-8)
 end
+
+
+@testset "remezord" begin
+    #
+    # Using the test-cases highlighted in [^Parks] Figures 8 and 15.
+    #
+    # [^Parks]: Rabiner, L. R., McClellan, J. H., & Parks, T. W. (1975). 
+    #   FIR digital filter design techniques using weighted Chebyshev 
+    #   approximation. Proceedings of the IEEE, 63(4), 595-610.
+    #
+    @test remezord(0.41665, 0.49417, 0.0116, 0.0001) == 39
+    @test remezord(0.135, 0.203, 0.1, 0.1) == 10
+    @test remezord(0.203, 0.135, 0.1, 0.1) == 10 # flipped HPF case.
+    @test remezord(0.00963, 0.13271, 0.0116, 0.0001) == 24
```

Comment:
I've tried actually designing these filters with `remez` and found that I needed minimum filter lengths 36, 14, 14, and 29 to fulfill the specs. That's more than the "1.3% error" promised in the docstring. Any idea why that is?

Also, the docstring only talks about the lowpass case, but here, we test for highpass, too.

Reply:
The function implementation I'm providing in this commit is based on Equation(s) 3.142 through 3.145 in Rabiner & Gold Theory and application of Digital Signal Processing. Rabiner/Gold cite the original paper in the Appendix section, namely the paper in the function header, which uses the same coefficients. The original paper claimed "%1.3 percent error", but I think the textbook makes a better statement on the estimation:

> The initial value of N along with the specified values Fp, Fs and d1/d2 are the initial guess as to the appropriate design parameters. Depending on whether the actual values of d1 and d2 are too high or too low, the value of N should be raised or lowered accordingly until specifications are just met or exceeded.

"d1" and "d2" in this case refer to the passband and stopband attenuation. Could you provide what calls you made to remez, and have you done a comparison with `firpmord` in MATLAB to see if you received the same minimum filter orders?

Reply:
As for the lowpass vs. highpass case, the design equation is based on the absolute frequency delta of the pass/stopbands, so I guess this is just a dangling test case we could eliminate. 

I appreciate the feedback you've provided so far, I don't have a full understanding or the context on what the roadmap for the DSP.jl repository should be.  I'm not looking to make the package more complicated. I hope you see that's the case as well, feel free to make suggestions if you think I'm not providing adequate code or good enough reasoning.

Reply:
Ok, maybe replace the 1.3% claim then with a more general warning that this is just an estimate?

This is what I did (and yes, I have to correct myself, for the case, even length 35 is sufficient):
```julia
julia> function testfilt(L, Wp, Ws, Rp, Rs)
           h = remez(L, [0, Wp, Ws, 0.5], [1, 0]; weight = 1 ./ [Rp, Rs], grid_density = 32)
           ω = range(0, π, length = 65536)
           H = freqresp(PolynomialRatio(h, [1]), ω)
           ip = findlast(<=(2π*Wp), ω)
           is = findfirst(>=(2π*Ws), ω)
           return maximum(x -> abs(abs(x) - 1), H[1:ip]) <= Rp && maximum(abs, H[is:end]) <= Rs
       end
testfilt (generic function with 1 method)

julia> testfilt(34, 0.41665, 0.49417, 0.0116, 0.0001)
false

julia> testfilt(35, 0.41665, 0.49417, 0.0116, 0.0001)
true

julia> testfilt(13, 0.135, 0.203, 0.1, 0.1)
false

julia> testfilt(14, 0.135, 0.203, 0.1, 0.1)
true

julia> testfilt(28, 0.00963, 0.13271, 0.0116, 0.0001)
false

julia> testfilt(29, 0.00963, 0.13271, 0.0116, 0.0001)
true

julia> function testfilt_hp(L, Wp, Ws, Rp, Rs)
           h = remez(L, [0, Ws, Wp, 0.5], [0, 1]; weight = 1 ./ [Rp, Rs], grid_density = 32, filter_type = isodd(L) ? filter_type_bandpass : filter_type_hilbert)
           ω = range(0, π, length = 65536)
           H = freqresp(PolynomialRatio(h, [1]), ω)
           ip = findlast(<=(2π*Ws), ω)
           is = findfirst(>=(2π*Wp), ω)
           return maximum(abs, H[1:ip]) <= Rp && maximum(x -> abs(abs(x) - 1), H[is:end]) <= Rs
       end
testfilt_hp (generic function with 1 method)

julia> testfilt_hp(13, 0.203, 0.135, 0.1, 0.1)
false

julia> testfilt_hp(14, 0.203, 0.135, 0.1, 0.1)
true
```
> what the roadmap for the DSP.jl repository should be

Yea, as if there was one...

Reply:
FWIW, I stumbled upon two formulas attributed to Bellanger and Kaiser, respectively, which give lengths 43, 10, 27 and 42, 9, 27 for these test cases. So that underlines that these are all estimates and ave to be treated as such.

So for me the take-away is that we should just reformulate the docstring to be more defensive.